feat: pytorch lightning adapter support

hamidzr opened the pull request:
## Description

- fix issue with hvd aggregation_frequency check in step_optimizer 05560c5
- add bactch_idx arg to evaluate_batch in pytrochTrial
- add PyTorch Lightning Adapter for Pytroch Trial
- add 2 usage examples

<!---
Lead with the intended commit body in this description field. For breaking changes,
please include "BREAKING CHANGE:" at the beginning of your commit body.
At minimum, this section should include a bracketed reference to the Jira ticket,
e.g. "[DET-1234]". When squash-and-merging, copy this directly into the description field.
-->


## Test Plan

<!---
Describe the situations in which you've tested your change, and/or a screenshot as appropriate.
Reviewers may ask questions of this test plan to ensure adequate manual coverage of changes.
-->


## Commentary (optional)

TODOs left
- [x] rename ptl to pl (pytorch_lightning's choice)
- [x] serialization issue when returning nested dict of metrics: this is inherited from PytorchTrial and expected
- [x] set the env image for examples once they're published on docker-hub
- [ ] remove dev tools from example

<!---
Use this section of your description to add context to the PR. Could be for particularly
tricky bits of code that could use extra scrutiny, historical context useful for reviewers, etc.
You may intentionally leave this section blank and remove the title.
--->


## Checklist

- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.


<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:
- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:
- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:
- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")
-->


[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

hamidzr commented on harness/determined/pytorch/_pytorch_context.py:
```diff
@@ -440,6 +440,7 @@ def step_optimizer(
         optimizer: torch.optim.Optimizer,  # type: ignore
         clip_grads: Optional[Callable[[Iterator], None]] = None,
         auto_zero_grads: bool = True,
+        on_before_zero_grad: Optional[Callable[[torch.optim.Optimizer], None]] = None,
```
this wouldn't be a breaking change would it?

hamidzr replied:
hmm why do we have the auto_zero_grads param then? 

also separately isn't something wrong with this check? 
```
        check.true(
            auto_zero_grads or self.hvd_config.aggregation_frequency > 1,
            "if optimizations.aggregation_frequency is larger than 1, "
            "you can only set auto_zero_grads to be true. ",
        )
```

shiyuann replied:
blocking: Actually, I take the word back. We shouldn't add this callback. The Pytorch interface is meant to let users call anything they want directly in `train_batch` rather than using callbacks. You should remove this callback.

`auto_zero_grads` is just for users who want to call zero grading themselves.

>  also separately isn't something wrong with this check?

Do you mean the check itself is wrong or the message?

hamidzr replied:
I removed it in the last review, there are no `auto_zero_grads` callbacks.


> Do you mean the check itself is wrong or the message?

￼the message? I don't know enough about hvd_config.aggregation_frequency to know for sure.

hamidzr replied:
TODO remove the callback from the argument list

hamidzr replied:
I included the fix for the aggregation frequency here.

hamidzr commented on examples/gan/_gan_mnist_pl/pl.py:
```diff
@@ -0,0 +1,262 @@
+            tqdm_dict = {'g_loss': g_loss}
+            output = OrderedDict({
+                'loss': g_loss,
+                # 'progress_bar': tqdm_dict, #  FIXME unserializable object torch.Tensor
```
```
 File "/run/determined/pythonuserbase/lib/python3.6/site-packages/determined/util.py", line 150, in json_serializer
<[]>
raise TypeError("Unserializable object {} of type {}".format(obj, type(obj)))
```

hamidzr replied:
I found in pytorch trial we don't support returning nested dictionaries in metrics yet.

shiyuann commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,209 @@
+        self, batch: TorchData, epoch_idx: int, batch_idx: int
+    ) -> Union[torch.Tensor, Dict[str, Any]]:
+        self.lm.on_train_batch_start(batch, batch_idx, dataloader_idx=0)
+        type(self.lm).global_step = mock.PropertyMock(return_value=batch_idx)
```
I tried `type(self.lm).global_step=batch_idx`. It worked.
 maybe we don't need this mock package?

shiyuann commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,209 @@
+            return _optimizers
+
+        for opt_idx, opt in enumerate(self.optimizers):
+            with monkey_patch.monkey_patch(self.lm, "optimizers", get_optimizers):
```
You can directly use lambda here.

shiyuann commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,209 @@
+
+        for opt_idx, opt in enumerate(self.optimizers):
+            with monkey_patch.monkey_patch(self.lm, "optimizers", get_optimizers):
+                self.lm.toggle_optimizer(opt, opt_idx)
```
I checked the source code of `toggle_optimizer`. It doesn't use `.optimizers` at all. Why do you need to monkey-patch here?

hamidzr replied:
1.1.8. it looks like they added it in 1.1.6

hamidzr replied:
what version did we use in prd?

hamidzr replied:
probably a version mismatch? it uses it here `0b7f5a88a pytorch_lightning/core/lightning.py          1183 (ananthsub               2021-02-04 14:50:57 -0800 1183)         for opt in self.optimizers(use_pl_optimizer=False):`

shiyuann commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,209 @@
+        for opt_idx, opt in enumerate(self.optimizers):
+            with monkey_patch.monkey_patch(self.lm, "optimizers", get_optimizers):
+                self.lm.toggle_optimizer(opt, opt_idx)
+            metrics = self.lm.training_step(batch, batch_idx, optimizer_idx=opt_idx)
```
Is it intentional to only make `optimizer_idx` keyword argument?

hamidzr replied:
I didn't actually give it much thought. is there a best practice around this that I should follow?

shiyuann replied:
I mean: why not just make all the arguments keyword arguments?

rb-determined-ai replied:
Isn't it unsafe to assume the user can accept `optimizer_idx`?  That's what I understand from the docs

hamidzr replied:
I used what the PL trainer does. https://github.com/determined-ai/determined/pull/1959/commits/11f1a4a659678a59d37c8c08a484ca98aff6feca

shiyuann commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,209 @@
+
+            opt_metrics.append(metrics)
+            with monkey_patch.monkey_patch(self.lm, "optimizers", get_optimizers):
+                self.lm.untoggle_optimizer(opt_idx)
```
I didn't see `LightningModule` had such member function `untoggle_optimizer`. Which version are you on?

hamidzr replied:
1.1.8. it looks like they added it in 1.1.6

shiyuann commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,209 @@
+        agg_metrics = {}
+        for opt_idx, rv in enumerate(opt_metrics):
+            for k, v in rv.items():
+                agg_metrics[f"opt{opt_idx}_{k}"] = v
```
I think it's wrong to add prefixes here since users are the ones to control the names. If they specify the same names for different optimizers, it should be wrong unless PytorchLightning itself implements in this way

shiyuann replied:
Actually, I am wrong. You need to add prefixes.

hamidzr replied:
I think we can probably keep the user's names unless they collide with each other. always adding prefixes was an easier first approach.

shiyuann commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,209 @@
+
+    def evaluate_batch(self, batch: TorchData) -> Dict[str, Any]:
+        batch_idx = self.context.current_batch()
+        type(self.lm).global_step = mock.PropertyMock(return_value=batch_idx)
```
Why do you need to set gloabl_step in validation? Is it a behavior of Pytorch Lightning?

hamidzr replied:
you're right we don't need to update it
`ag 'global_step.* = ' `

hamidzr replied:
I was getting the batch index here wrong. it should be batch index scoped to the epoch. I might need to add batch_idx as an arg to `evaluate_batch` to pass it through

hamidzr replied:
I added this argument and a patch for it for backward compat. can you take a look, please?

shiyuann replied:
>  I might need to add batch_idx as an arg to evaluate_batch to pass it through

You should instead add a function in the context object `get_cur_batch_idx` rather than change the arguments. We are moving towards using the context object as many times as possible.

hamidzr replied:
FTR we decided to keep the batch_idx argument.

shiyuann commented on examples/gan/_gan_mnist_pl/model_def.py:
```diff
@@ -0,0 +1,30 @@
+                    b2=context.get_hparam('b2'),
+        )
+
+        context.wrap_model(lm.generator.model)
```
shouldn't this be called within `PTLAdapter.__init__`?

hamidzr replied:
I was thinking we leave this out for the user to call for multi model/optimizer algorithms like GANs

shiyuann replied:
You could actually just call wrap_model on the LightningModule

hamidzr replied:
wouldn't wrap model need to be called once per nn module?

shiyuann replied:
yeah all the submodules will be called on automatically. LightningModule itself is a nn.Module.

hamidzr replied:
I moved this into `__PLAdapter.__init__`  we wrap the passed in lightning module for the user and relying on the pytorch's behavior on sub model for nn.Module. If the user has separate models defined that are not part of the lM they'll have to call wrap_model on it themselves in `model_def`. does that sound right?

shiyuann replied:
> If the user has separate models defined that are not part of the lM they'll have to call wrap_model on it themselves in model_def. does that sound right?

Yes if they don't define the models to be a member of the lightning module.

shiyuann commented on harness/determined/errors.py:
```diff
@@ -72,3 +72,12 @@ class SkipWorkloadException(Exception):
     pass
+
+
+class UnsupportedException(Exception):
```
`UnsupportedException` is a pretty vague name. I'd use 
```
class InvalidModelException(InvalidExperimentException):
```

shiyuann replied:
I didn't see this file updated.

hamidzr replied:
I must have missed it. updated.

shiyuann commented on harness/determined/pytorch/_lightning/README.md:
```diff
@@ -0,0 +1,168 @@
+# PTL Adapter
```
Is this `README.md` ready to be merged? It looks more like a development note rather than being useful to users.

hamidzr replied:
no, that's right. it's here mostly for reviewers benefit I'll take it out

shiyuann reviewed (commented):
Is this PR in a ready-to-be-merged state? I saw two examples, which I assume will only leave the GAN in the codebase. I also saw a `README.md` that seems not to be for users. Otherwise, the PR looks fine overall. There are places that are not consistent with our coding style in other harness code.

hamidzr commented on examples/computer_vision/_mnist_ptl/.detignore:
```diff
@@ -0,0 +1 @@
+lightning_logs/
```
if we don't ignore it they'll get uploaded to determined with the experiment and show up as dirty files in git (at least for how the experiment is set up right now) do you see any downsides for having it?

shiyuann replied:
The downside is that it introduces an unnecessary file to the examples. You should just set the logs to use `/tmp`.

hamidzr commented on examples/gan/_gan_mnist_pl/model_def.py:
```diff
@@ -0,0 +1,30 @@
+from determined.pytorch import PyTorchTrial, PyTorchTrialContext, DataLoader as DetDataLoader
```
I found just `DataLoader`  to be a bit confusing for myself but that's subjective. I'll update it to be consistent.

hamidzr commented on examples/gan/_gan_mnist_pl/model_def.py:
```diff
@@ -0,0 +1,30 @@
+from determined.pytorch._lightning import PTLAdapter
+import pl
+
+class GanTrial(PTLAdapter):
```
I might be missing something here wdy mean? it is `GanTrial` already

shiyuann replied:
Sorry. I mistyped. It should be `GANTrial`. `GAN` should use all capital characters.

hamidzr commented on harness/determined/pytorch/_lightning/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from determined.pytorch._lightning.adapter import (
+    HyperparamsProvider,
```
I tried two different ways of passing the context down in the two examples the gan one doesn't use this but the other one does. IMO it's helpful for usecase 1 but the GAN example showcases PL's use of `.hparams` and `save_hyperparameters` which if the user opts to use they wouldn't need this type definition.

hamidzr replied:
I'll just keep the more native pl usage which also means removing the `HyperparametersProvider`

hamidzr commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,209 @@
+    return kw_present
+
+
+def check_compat(lm: ptl.LightningModule):
```
yes to check the compatiblity of LM with PTLAdapter. any name suggestions?

shiyuann replied:
Maybe `check_compatability`?

hamidzr commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,209 @@
+        self.optimizers = optimizers
+
+        # user should merge these with their own callbacks
+        self.build_callbacks = lambda: build_callbacks(lightning_module, self.context)
```
good point. I updated it to be how pytorch trial sets it up. would it be reasonable to expect the users to call `super().build_callbacks` and merge the output with their callbacks. or should I try to do that for them maybe during `__init__`?

shiyuann replied:
Yes. I think it's fine to let users call `super().build_callbacks` as long as we are clear of it in the docstring.

hamidzr replied:
I updated the docstring

hamidzr commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,209 @@
+    raise UnsupportedException(msg)
+
+
+def expects_param(fn: Callable[..., Any], pos: Optional[int], name: str) -> bool:
```
👍🏼 I updated it to has_param. do you have any thoughts suggestions on the question I left here in the code?

shiyuann replied:
`has_param` looks like it should be in `determined/util.py` together with `is_overridden`. What question? I didn't see it.

hamidzr replied:
👍🏼 . nvm I removed the question

hamidzr commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,209 @@
+
+
+class PTLAdapter(PyTorchTrial):
+    context: PyTorchTrialContext
```
I thought this is a hint for the typesystem but I might be wrong.

shiyuann replied:
Okay. It looks fine.

rb-determined-ai replied:
Wait, PyTorchTrial is for the user to define, why are we putting annotations on user objects?

shiyuann replied:
`PTLAdapter` is a middle layer between user-defined and training-internal "layers". I see this class as something that can be user-defined.

hamidzr commented on harness/determined/pytorch/_callback.py:
```diff
@@ -72,3 +72,33 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
         """
         pass
+
+    def on_before_zero_grad(self) -> None:
```
leftover. I removed it

hamidzr commented on examples/computer_vision/_mnist_ptl/.gitignore:
```diff
@@ -0,0 +1 @@
+lightning_logs/
```
let's chat on the other comment. https://github.com/determined-ai/determined/pull/1959#discussion_r579528864

hamidzr commented on examples/gan/_gan_mnist_pl/.detignore:
```diff
@@ -0,0 +1 @@
+lightning_logs/
```
let's chat on the other comment. #1959 (comment)

hamidzr commented on examples/computer_vision/_mnist_ptl/ptl.py:
```diff
@@ -0,0 +1,105 @@
+import torch
```
hmm why not both? this one is much simpler could it be useful?

hamidzr commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -20,6 +20,16 @@
+    """
+    A helper to iterate over a dataset using start and end indices.
+    """
+    for epoch_idx in range(start // dataset_len, end // dataset_len + 1):
```
we can't really use `self` considering this being defined as a utility function we could maybe add this to the trial controller but instead, IMO it might be better to name that division to make it more explicit? I update it.

shiyuann replied:
There are two issues with this function. One issue is that this function is introducing the dependency of our training loop on the concept of the epoch, which is dependent on the length of the training loader. If the training loader returns a wrong number, the current training loop still works except for the epoch-related callbacks. The other issue is that this function is duplicate logic from `PyTorchTrialController.get_epoch_idx`. I meant to just use `self.context. get_epoch_idx(idx)` directly in the loop if you really need to use it. Otherwise, I don't see any benefits to introduce this complexity of writing two-level loops.

hamidzr replied:
the duplicate logic issue for `get_epoch_idx` we can solve by another utility fn but it's small enough that I didn't. 

hmm one benefit that breaking up the loop will provide would be that it'll let us do operations we need to do on epoch boundaries without having to call context.is_epoch_end every iteration. for example we need it to split per_batch_metrics by epoch for on_train_epoch_end callback. that overhead seems small.

do we support dataloaders where we can't know the length of or with variable lengths?

hamidzr replied:
I reverted back to the old loop

hamidzr commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,209 @@
+
+class PTLAdapter(PyTorchTrial):
+    context: PyTorchTrialContext
+    lm: ptl.LightningModule
```
same as https://github.com/determined-ai/determined/pull/1959#discussion_r579552501 let's chat there

hamidzr commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,209 @@
+                'strict': True,  # enforce that the monitor exists for ReduceLROnPlateau
+            }
+            """
+            # TODO support custom frequencies with the manual step.
```
DET-5021

hamidzr commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -242,42 +252,51 @@ def _train_for_step(
+                    tr_metrics[name] = metric
+
+                epoch_per_batch_metrics.append(tr_metrics)
+            for callback in self.callbacks.values():
```
good catch. I assume you mean `on_train_epoch_end`?

hamidzr commented:
thanks for the review. I addressed the comments. For some of them I left more comments that'd need your attention.

I'm working on addressing the mypy lint failures. some of them are a bit unclear for example:
`determined/pytorch/_pytorch_context.py:443: error: Name 'torch.optim.Optimizer' is not defined` where it has no problem finding it for two lines above it.

shiyuann commented:
@hamidzr It seems using `# type: ignore` is fine.

shiyuann commented on examples/gan/_gan_mnist_pl/model_def.py:
```diff
@@ -0,0 +1,27 @@
+from determined.pytorch import PyTorchTrial, PyTorchTrialContext, DataLoader
+from determined.pytorch._lightning import PLAdapter
+import pl
```
Can you rename the file `pl.py` to be something else? maybe `gan.py`. It's a bit confusing.

hamidzr replied:
did the same for the other example as well

shiyuann commented on examples/gan/_gan_mnist_pl/pl.py:
```diff
@@ -0,0 +1,253 @@
+# -*- coding: utf-8 -*-
```
@rb-determined-ai If we include the original file from Pytorch Lightning, do we need to include their copyright info here?

shiyuann commented on harness/determined/monkey_patch.py:
```diff
@@ -30,3 +31,15 @@ def inner(*args, **kwargs):  # type: ignore
     return wrap
+
+
+def has_param(fn: Callable[..., Any], name: str, pos: Optional[int] = None) -> bool:
```
It seems that it should exist in `determined/util.py`

shiyuann commented on harness/determined/pytorch/_callback.py:
```diff
@@ -72,3 +72,27 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        """
+        pass
+
+    def on_validation_epoch_start(self) -> None:
```
Use `on_validate_epoch_start` to be consistent with the other names?

hamidzr replied:
hmm what are you suggesting to be consistent with? our other callbacks start with `on_validation`
```
hrz ➜ determined git:(ptl-2) ag on_valida* harness/determined/pytorch/_callback.py                                                                                                   
17:        (except for :meth:`on_validation_end`, :meth:`on_validation_step_end` and
22:    def on_validation_start(self) -> None:
28:    def on_validation_end(self, metrics: Dict[str, Any]) -> None:
37:    def on_validation_step_start(self) -> None:
44:    def on_validation_step_end(self, metrics: Dict[str, Any]) -> None:
88:    def on_validation_epoch_start(self) -> None:
94:    def on_validation_epoch_end(self, outputs: List[Any]) -> None:
```

hamidzr replied:
or on.train to on.training

hamidzr replied:
I updated ontrain to be consistent with the rest. let's continue in slack with @ahj-determined

shiyuann commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,184 @@
+
+
+def check_compat(lm: pl.LightningModule) -> None:
+    prefix = "Unsupported usage in PLAdapter: "
```
```suggestion
    prefix = "Unsupported usage in PTLAdapter: "
```

hamidzr replied:
hmm  I changed these when we talked about using PL instead of PTL everywhere

shiyuann commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,184 @@
+        def on_validation_epoch_start() -> None:
+            lm.on_validation_epoch_start()
+
+        cb.on_validation_epoch_start = on_validation_epoch_start
```
Dynamically changing a member field of a Python object is generally a bad practice, which makes code hard to debug. Why don't you just define an inner class:

```
class PTLAdapterCallback(PyTorchCallback):
   def ...():
      ...
```

shiyuann commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,184 @@
+                'strict': True,  # enforce that the monitor exists for ReduceLROnPlateau
+            }
+            """
+            # TODO(DET-5021) support custom frequencies with the manual step.
```
What are custom frequncies?

hamidzr replied:
interval != 1

shiyuann commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,184 @@
+        self, batch: TorchData, epoch_idx: int, batch_idx: int
+    ) -> Union[torch.Tensor, Dict[str, Any]]:
+        self.lm.on_train_batch_start(batch, batch_idx, dataloader_idx=0)
+        type(self.lm).global_step = batch_idx
```
Is it set after calling `on_train_batch_start` in Lightning?

hamidzr replied:
it should be set before all the hooks.

hamidzr replied:
it starts at 0 and then increments https://github.com/PyTorchLightning/pytorch-lightning/blob/6cc1a060785448eacd016f6689d592ac822adaaf/pytorch_lightning/trainer/training_loop.py#L552

shiyuann commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,184 @@
+                self.lm.toggle_optimizer(opt, opt_idx)
+            metrics = self.lm.training_step(batch, batch_idx, optimizer_idx=opt_idx)
+
+            self.lm.on_train_batch_end(metrics, batch, batch_idx, dataloader_idx=0)
```
Should this be called after the backward pass?

hamidzr replied:
yeah you're right https://github.com/PyTorchLightning/pytorch-lightning/blob/6cc1a060785448eacd016f6689d592ac822adaaf/pytorch_lightning/trainer/training_loop.py#L505

hamidzr commented on examples/computer_vision/_mnist_pl/model_def.py:
```diff
@@ -0,0 +1,31 @@
+
+from determined.pytorch import PyTorchTrial, PyTorchTrialContext, DataLoader
+from determined.pytorch._lightning import PLAdapter
+import pl
```
this one is just a relative import. maybe I should rename the file

rb-determined-ai commented on examples/computer_vision/_mnist_pl/model_def.py:
```diff
@@ -0,0 +1,31 @@
+"""
```
This example will not be part of the PR that lands, right?  I'm guessing that's why `_mnist_py` starts with a `_`

hamidzr replied:
it is why? depending on when we land it it might initially be with the _

shiyuann replied:
@rb-determined-ai The example will land as part of the PR. And let's remove the underscore in the file names.

rb-determined-ai reviewed (commented):
I'm mostly reviewing the API here.

Overall, this is quite a bit harder of a challeng than I was thinking, and I think you have a great start with a ton of details already covered.

I think @shiyuann's suggesting of subclasses here could result in a cleaner API.  The attributes that we want to overwrite on the lightning module could be set more cleanly.

Also, even though I normally really am opposed to mixing user code and Determined code, and even though I'm normally opposed to APIs where `super().__init__()` is part of the API, I agree with your design; those are the right choices in this case.  But we can make things a bit better by namspacing the variables we use into a plain-old-data class.

I'm thinking something like:
```python
# namespaced variables
class _LightningAdapterState:
    def __init__(self, context, lm, ...):
        self.context = context
        self.lm = lm
        ...

# lightweight extension of PyTorchTrialContext
class PyTorchLightningContext(PyTorchTrialContext):
    def __init__(self, ...):
        super().__init__(...)
        self._lm = None

    # user-facing call
    def wrap_lightning_module(self, lm):
        assert self._lm is None, "don't call this twice"

        # dynamically-created class, bound around variable `lm`
        class _WrapperClass(type(lm)):
            def __init__(self):
                self.ddp = False
                self.dp = False
                ...

            def __getattr__(wrapper, name):
                return getattr(lm, name)

            def __setattr__(wrapper, name, value):
                return setattr(lm, name, value)

            def __delattr__(wrapper, name):
                return delattr(lm, name)

            def optimizers(wrapper, use_pl_optimizer: bool = False):
                assert use_pl_optimizer is False, "use_pl_optimizer not supported"
                # do some determined magic?
                ...
                return lm.optimizers()

        return _WrapperClass()

class PyTorchLightningTrial(PyTorchTrial):
    trial_context_class = PyTorchLightningContext

    def __init__(self, context):
        if context._lm is None:
            raise ValueError("wrap before super().__init__ please")

        # low-impact on user namespace
        self._adapter_state = _LightningAdapterState(
            context = context,
            lm = context._lm,
            ...
        )

    def build_callbacks(self):
        # based on self._adapter_state
        ...

## user code

class MyTrial(PyTorchLightningTrial):
    def __init__(self, context):
        self.context = context
        self.my_lm = ...
        self.my_lm = context.wrap_lightning_module(my_lm)
        super().__init__(context)
```

shiyuann commented:
@rb-determined-ai I am not opposed to separate out user code and Determined code. But having a class that has only abstract methods and no member variables is an extreme, IMO. Generally, the design in Keras and Pytorch uses inheritance, which allows users to put everything they define in a separate class but have good interaction with the library code. I have also seen `super().__init__` being part of the API design many times. That being said, I think it's unnecessary to have a dynamically generated class just to provide access to the context object. Also, I think generally using a dynamically generated class is bad, which really abuses Python and makes code hard to read and debug. (I think the Keras interface needs to be changed.)

Besides, I think having a separate namespace is contradictory to the goal of the adapter approach. `PTLAdapter` aims at providing a middle layer between `PytorchTrial` and user code. `PyTorchLightningTrial` and `PyTorchLightningContext` sounds more like providing a separate training loop context by hiding the details with inheritance, which also will let users mix with the extension support.

rb-determined-ai commented:
> Besides, I think having a separate namespace is contradictory to the goal of the adapter approach

How so?  Users who use `PyTorchLightningTrial` do not want to learn the innards of `PyTorchLightningTrial`, and they won't want name conflicts between us and them.  Those sorts issues are very hard to debug for users because from the user's perspective, there is absolutely no reason why their code is wrong; it just breaks after the `super().__init__()` call because they also set `self.lm` or something.

(I feel pretty strongly this namespacing of our variables; I don't care nearly as much about the dynamic class vs monkeypatching thing)

> But having a class that has only abstract methods and no member variables is an extreme

What do you mean?  Neither class I sketched out is like that.

> it's unnecessary to have a dynamically generated class just to provide access to the context object. Also, generally using a dynamically generated class abuses Python and makes code hard to read and debug

What do you mean?  The dynamically generated class is for overriding attributes and methods without monkeypatching.  I think the monkeypatching strategy is bad because it looks to me like we are going to be making assumptions about *when* the user is going to call e.g. `optimizers`.

> Also, generally using a dynamically generated class abuses Python and makes code hard to read and debug

I mostly agree with you, I just think it's less bad than monkeypatching.  But what is the real "right" answer?  I think the real right answer is to *stop trying to control the training loop*.  We should be allowing users to just invoke the real lightning trainer with a Determined callback.  After having looked closer at the PTL training loop while reviewing this PR, I have real doubts that we can imitate the PTL trainer well enough to offer a good experience, and I would personally prefer that we focus on push architecture and sprinkle API first.

Side question: are there times that users have complained about readability or debuggability of TFKerasTrial due to the wrapped class?  I have consistently seen users complain about TFKerasTrial having different behavior than tf.keras, but I have not ever heard somebody complain about the dynamic class.

hamidzr commented on examples/computer_vision/_mnist_pl/model_def.py:
```diff
@@ -0,0 +1,31 @@
+"""
+This example shows how to interact with the Determined PyTorch interface to
```
how about 

"""
This example shows how to interact with the Determined PyTorch Lightning Adapter
interface to build a basic MNIST network. The PLAdapter utilizes the provided
LightningModule with Determined's trainer.
"""

hamidzr replied:
Determined's trainer. => determined's pytorch control loop

hamidzr commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,184 @@
+        self.lm.to(self.context.device)
+
+    def build_callbacks(self) -> Dict[str, PyTorchCallback]:
+        context = self.context
```
less verbosity. it's copying only a reference right?

shiyuann commented:
@rb-determined-ai 
There are a couple of questions here:

1. adapter support vs. extension support

> After having looked closer at the PTL training loop while reviewing this PR, I have real doubts that we can imitate the PTL trainer well enough to offer a good experience, and I would personally prefer that we focus on push architecture and sprinkle API first.

This is already been discussed at the beginning of the project. Although we couldn't avoid imitating Trainer, the places we do so is where we set the epoch index, batch index, and etc. The pro is building this adapter support can get immediate benefits to users who want to use LightningModule and don't care about Trainer a lot. The decision made was to use a couple of weeks to build this and make users happy.


> I think the real right answer is to stop trying to control the training loop. We should be allowing users to just invoke the real lightning trainer with a Determined callback.

yeah. What you described is the extension support.

2. monkey patching vs dynamic subclassing

Essentially, both of these two methods are imitating the trainer. They are both bad in dynamically changing the original methods or variables. But the reason I dislike dynamic subclassing is that it creates an unnecessary type.

3. Namespacing (PyTorchLightningTrial) vs. no separate namespace

`LightningTrial` is already booked for the extension support. If we provide two trials, it will be confusing to understand two different types of trials. However, if we provide `PTLAdapter(PytorchTrial)` and `LightningTrial`, it would be very clear that `PTLAdapter` is based on our Pytorch training loop and users shouldn't expect it behaves the same as Lightning itself without Determined.

4. getting rid of super().init() in the API

This is actually very common practice, at least being used widely in Pytorch Lightning. So I think it's totally okay to have such design for Lightning users.

5. Side question: are there times that users have complained about readability or debuggability of TFKerasTrial due to the wrapped class? I have consistently seen users complain about TFKerasTrial having different behavior than tf.keras, but I have not ever heard somebody complain about the dynamic class.

We don't. I think generally we should avoid both monkey patching and dynamic subclassing. The submodeling in tf.keras was introduced by the Native API, which tries to be accurate and flexible while fit into the `TFKerasTrialController`. It is a bit unnecessary if we remove the Native API. I am more in favor of the following design

```
class TFKerasTrial:
def __init__():
    self.model = self.context.Model(...)
    self.context.compile(...)

def train():
    self.model.fit()

def validate():
    self.model.evaluate()
```

shiyuann commented:
@rb-determined-ai feel free to join our weekly sync if you think we need to change anything.

hamidzr commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,214 @@
+        context.wrap_model(lightning_module)
+        optimizers, lr_schedulers = self.setup_optimizers_schedulers(context, lightning_module)
+        pls = _PLAdapterState(context, lightning_module, optimizers)
+        self._pls = pls
```
short name suggestions with low collision possibility?
pls: pl state

shiyuann replied:
just `_pl`? I am fine with any names.

shiyuann commented on examples/computer_vision/_mnist_pl/adaptive.yaml:
```diff
@@ -0,0 +1,17 @@
+description: mnist_pytorch_lightning_adaptive
+max_restarts: 1
```
remove this when merging

shiyuann commented on harness/determined/pytorch/_pytorch_trial.py:
```diff
@@ -308,8 +313,14 @@ def _train_for_step(
+                per_batch_metrics += epoch_per_batch_metrics
+                epoch_per_batch_metrics = []
 
+        per_batch_metrics += epoch_per_batch_metrics
```
This isn't right. If a step has batches in two sequential epochs, `per_batch_metrics` contains the metrics of all the batches but `epoch_per_batch_metrics` would clear itself once it hits an epoch end.

hamidzr replied:
👍🏼  we removed support for on_train_epoch_end since we can't guarantee epochwide metrics for this callback with mid-epoch preemption

shiyuann commented on examples/computer_vision/_mnist_pl/model_def.py:
```diff
@@ -0,0 +1,26 @@
+        super().__init__(context, lightning_module=lm)
+
+    def build_training_data_loader(self) -> DataLoader:
+        self.dm.setup()
```
This doesn't seem to be right since this function downloads data to the same directory for all the GPUs, which might cause errors

hamidzr replied:
I moved the download section to dm.prepare() as PL suggests and used local_rank to point it to avoid collisions

shiyuann commented on examples/gan/_gan_mnist_pl/model_def.py:
```diff
@@ -0,0 +1,33 @@
+        self.dm.prepare_data()
+
+    def build_training_data_loader(self) -> DataLoader:
+        self.dm.setup()
```
same

hamidzr commented on examples/computer_vision/_mnist_pl/mnist.py:
```diff
@@ -0,0 +1,94 @@
+    def __init__(self, *args, lr: float, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.save_hyperparameters()
```
So that the inputs get saved under `self.hparams` if I'm not reading this wrong https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#save-hyperparameters

hamidzr commented on harness/determined/util.py:
```diff
@@ -37,6 +38,18 @@ def is_overridden(full_method: Any, parent_class: Any) -> bool:
     return False
 
 
+def has_param(fn: Callable[..., Any], name: str, pos: Optional[int] = None) -> bool:
```
that was my question too. Would param fit better if we go by this? https://stackoverflow.com/a/1788926

hamidzr commented:
I squashed the history so that the reviews are easier to read.

hamidzr commented:
I did some testing with slots_per = 2 on a single machine/agent. it might be good to also test in a multi-agent scenario

hamidzr commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,275 @@
+import inspect
```
should this file be named _adapter?

shiyuann commented on harness/determined/pytorch/_lightning/adapter.py:
```diff
@@ -0,0 +1,275 @@
+
+        def lightning_scheduler_dict_to_det(lrs: dict) -> _LRScheduler:
+            """
+            input_dict = {
```
blocking: Need to add checks for the fields that we don't support. We need to support frequency. If `monitor` and `strict` are set, we should add checks and fail.

hamidzr replied:
> We need to support frequency.

let's leave this for a follow-up pr. I'll add the checks

hamidzr replied:
https://determinedai.atlassian.net/browse/DET-5021

shiyuann commented on examples/gan/_gan_mnist_pl/distributed.yaml:
```diff
@@ -0,0 +1,29 @@
+description: gan_mnist_pytorch_lightning_adaptive
```
```suggestion
description: gan_mnist_pytorch_lightning_distributed
```

shiyuann reviewed (approved):
good work!

hamidzr commented on examples/computer_vision/_mnist_pl/pl.py:
```diff
@@ -0,0 +1,96 @@
+
+class MNISTDataModule(pl.LightningDataModule):
+    def __init__(self):
+        super().__init__()
```
Just following the pattern that lightning suggests https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#minimal-example

hamidzr commented:
thanks for the many reviews!
